python/lsst/daf/butler/registry/interfaces/_database.py

```diff
@@ -335,31 +337,86 @@ def fromConnection(cls, connection: sqlalchemy.engine.Connection, *, origin: int
+            A list of tables to lock for the duration of this transaction.
+            These locks are guaranteed to prevent concurrent writes, but only
+            prevent concurrent reads if the database engine requires that in
+            order to block concurrent writes.
+
+        Notes
+        -----
+        All transactions on a connection managed by one or more `Database`
+        instances _must_ go through this method, or transaction state will not
+        be correctly managed.
         """
         assert not (interrupting and self._connection.in_transaction()), (
             "Logic error in transaction nesting: an operation that would "
             "interrupt the active transaction context has been requested."
         )
-        if self._connection.in_transaction():
+        # We remember whether we are already in a SAVEPOINT transaction via the
+        # connection object's 'info' dict, which is explicitly for user
+        # information like this.  This is safer than a regular `Database`
+        # instance attribute, because it guards against multiple `Database`
+        # instances sharing the same connection.  The need to use our own flag
+        # here to track whether we're in a nested transaction should go away in
+        # SQLAlchemy 1.4, which seems to have a
+        # `Connection.in_nested_transaction()` method.
+        savepoint = savepoint or self._connection.info.get(_IN_SAVEPOINT_TRANSACTION, False)
+        self._connection.info[_IN_SAVEPOINT_TRANSACTION] = savepoint
+        if self._connection.in_transaction() and savepoint:
             trans = self._connection.begin_nested()
         else:
-            # Use a regular (non-savepoint) transaction only for the outermost
-            # context.
+            # Use a regular (non-savepoint) transaction always for the
+            # outermost context.
             trans = self._connection.begin()
+        self._lockTables(lock)
         try:
             yield
             trans.commit()
         except BaseException:
             trans.rollback()
             raise
+        finally:
+            if not self._connection.in_transaction():
+                self._connection.info.pop(_IN_SAVEPOINT_TRANSACTION, None)
+
+    @abstractmethod
+    def _lockTables(self, tables: Iterable[sqlalchemy.schema.Table] = ()) -> None:
+        """Acquire locks on the given tables.
+
+        This is an implementation hook for subclasses, called by `transaction`.
+        It should not be called directly by other code.
+
+        Parameters
+        ----------
+        tables : `Iterable` [ `sqlalchemy.schema.Table` ], optional
+            A list of tables to lock for the duration of this transaction.
+            These locks are guaranteed to prevent concurrent writes, but only
+            prevent concurrent reads if the database engine requires that in
+            order to block concurrent writes.
+        """
+        raise NotImplementedError()
```

Comment:
General comment about table locking - it is not a standard SQL concept and implementation across backends can be inconsistent. MySQL in particular (I'm not sure if we care about it much) has a nasty interaction between transactions and LOCK TABLES. In some cases it may be possible to use SERIALIZABLE transaction isolation to the same effect as LOCK TABLE, but that is case by case consideration. And with LOCK TABLE things are usually much faster for the client that can lock tables but it can also slow down all other concurrent clients.

Reply:
I certainly plan to use this as little as possible, but it seems to be the only way to make certain multi-statement operations atomic while guaranteeing invariants that can't be written as constraints.

My hope was that the very specific and limited guarantees made by this interface would be something all backends could be expected to support.  It sounds like that's at least plausible for MySQL, even if it might be a pain, and that's good enough for me right now.